Turbomole model

sriz1 opened the pull request:
Turbomole interface with mudslide

Commit:
changed model.py and added some constants file

Commit:
models.py is edited

Commit:
test_turbomole.py file was added

Commit:
models.py generated with dos2unix

Commit:
minor changes to adjust changes in parsers

smparker commented on mudslide/amu_to_au.py:
```diff
@@ -0,0 +1 @@
+amu_to_au = 1822.88
```
Is this the NIST value? I'm sure it has more digits than this...

smparker replied:
Also, this is a constant, so should be put in the `constants.py` file.

sriz1 replied:
All the constants are now under constants.py file. Missing file periodic_table.py is now added.

smparker commented on mudslide/constants.py:
```diff
@@ -3,3 +3,4 @@
 
 boltzmann = 3.166811563455557e-6
 eVtoHartree = 27.211386245988
+temperature = 300.0
```
This is not a constant. Remove.

sriz1 replied:
done!

smparker commented on mudslide/models.py:
```diff
@@ -4,12 +4,21 @@
 from typing import Any
 from .typing import ArrayLike, DtypeLike
 from .constants import eVtoHartree
+from .periodic_table_atomic_mass import masses
```
This file has not been added to the repo. Also, a better name would just be `periodic_table`, so then you'd have `from .periodic_table import masses`, and it would eventually have a more natural way to also pull other data from the periodic table.

sriz1 replied:
done!

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
         return out.reshape([5, 2, 2])
 
 
+class TMModel(AdiabaticModel_):
```
Please move this class to a new file.

sriz1 replied:
TMModel is now a separate file which is NOT part of models.py ( I am assuming that was your suggestion).

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+        coords = []
+        self.atom_order = []
+
+        try:
```
This is too hardcoded. The solution for finding which file has the `$coord` data should be more flexible and will end up being simpler. It really should just be to look in control and find a (non-comment) line that starts with `$coord`. Then if the rest of the line includes `file=<something>`, then that's the file where you will find the coordinates.

sriz1 replied:
Addressed.

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+        self.atom_order = []
+
+        try:
+            read_start = np.argwhere(np.array(["$coord\n" in line for line in lines]))[0][0]
```
This is really wasteful. You're basically turning the whole thing into an array of `True`/`False` just to find the first line, and then only returning the line number. It will also fail if anything even slightly different is encounted. For example,
```
$coord ! this is an allowed comment by turbomole
```
would fail because there are things after the `d`.
```
 $coord
```
would fail because of the space. `$coord ` would fail because of the space at the end of the line. All of those are allowed by turbomole, so should be covered by us.

The best thing to do though is to use some built in facilities in turbomole. The `sdg` script/program will do most of what you need. For example,
```
sdg coord
```
will print out the coord block like
```
$coord
    2.68245142153191     -0.01532925226678      0.00000012933989  c
    5.35661968932518     -0.01570554780053     -0.00000007307270  c
   -1.37551527364443      2.32857147093840      0.00000011469638  c
   -2.71372989497353      4.64270033461115     -0.00000008888730  c
    5.16301691170948      8.97194268320440      0.00000011508832  c
    3.82944899733982      6.65643153409124     -0.00000006556916  c
   10.35925385641554     -0.01351664484525     -0.00000010603421  c
    7.68662189081884     -0.01396492275135      0.00000006587935  c
   -5.21925276592780      8.97196059766757     -0.00000010870875  c
   -3.88677960360603      6.65562142964378      0.00000006945337  c
    1.32870757012246      2.32575323529588     -0.00000011705168  c
    2.66497386930571      4.63899958287558      0.00000006486653  c
```
Running `sdg -f` will tell you just the filename that the data group is in. For example,
```
> sdg -f coord
coord
> sdg -f dft
control
```

Using the built-in functions is a much better idea because they are maintained by turbomole so they will better handle edge cases and things like that.

sriz1 replied:
Addressed!

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+        self.mass = np.array(self.mass, dtype=np.float64).reshape(self.ndim()) * amu_to_au
+
+
+    def update_coords(self, X):
```
This is really weirdly hardcoded. Rewrite this using the built in formatting features in python. It will be like 4 lines long if you do that.

sriz1 replied:
Used python built in formatting!

smparker replied:
Use fixed widths and the scientific notation to make sure you don't lose any numerical precision. I think something like `{:20.16g} {:20.16g} {:20.16g} {:3s}` should work well.

sriz1 replied:
updated!

smparker replied:
The most recent version of the code still uses the `float` format (`{:22.14f}`). That will lead to precision loss for small values. Use the suggested format above, which will be guaranteed to preserve the full precision of the coordinates. If turbomole complains about the `e` in the output, just use `line.replace("e", "d")` to fix it.

sriz1 replied:
OK now (?)  line 109 - 113 

<img width="941" alt="Screen Shot 2022-04-19 at 11 08 11 AM" src="https://user-images.githubusercontent.com/46786940/164035776-e3df81dd-3461-4802-8a19-29c18f15cefb.png">

smparker replied:
Use `16g`. To be on the safe side.

smparker replied:
Okay, after some testing, the better format string to use is `{: 22.16e}`.

sriz1 replied:
hmm that was not working and I think because of the space issue (no space between the coordinates) so I instead used line 112 
<img width="758" alt="Screen Shot 2022-04-19 at 2 04 38 PM" src="https://user-images.githubusercontent.com/46786940/164067520-904f658b-8ae9-4704-8136-fe626757bd35.png">

The coord now looks like this 

<img width="713" alt="Screen Shot 2022-04-19 at 2 05 19 PM" src="https://user-images.githubusercontent.com/46786940/164067613-e661b790-0708-4073-a089-49db39789faa.png">

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+    def turbomole_init(self):
+        self.coord_path = Path(self.turbomole_dir)/"control"  
+        self.get_coords()
+        self.call_and_parse_turbomole()
```
Why does the initialization calculate the first step? That's not in keeping with how the rest of the models are designed and I don't think it's a good design.

sriz1 replied:
resolved!

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+            coord_file.write("".join(lines))            
+
+
+    def call_and_parse_turbomole(self, outname="turbo.out"):
```
This needs a more descriptive name. For one, the parse part can be implied. But also, if you're giving it such a general name (`call`), then it should be flexible enough to handle just about any call you'd want to make. Right now it's hardcoded for `ridft`, `rdgrad`, and `egrad`. What if you wanted to customize that?

sriz1 replied:
If you could provide some details on what exactly we want to achieve - would be great! Thanks!

sriz1 replied:
Hopefully this is now addressed! now 
<img width="746" alt="Screen Shot 2022-04-10 at 2 36 43 PM" src="https://user-images.githubusercontent.com/46786940/162634443-0710578d-60be-428b-95d0-5066a17b2818.png">

<img width="748" alt="Screen Shot 2022-04-10 at 2 37 43 PM" src="https://user-images.githubusercontent.com/46786940/162634489-1ac7eb27-bc2b-4bf5-8bd1-10c4796504c6.png">

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+            for j in range(self.nstates()):
+                if i != j:
+                    for dct in parsed_nac_coupling:
+                        if (dct["bra_state"] == i) and (dct["ket_state"] == j):
```
This is a wasteful and convoluted way to go through all the data. A better looping structure would just be
```
for dct in parsed_nac_coupling:
    i = dct["bra_state"]
    j = dct["ket_state"]
    ...
```

Then you've removed two loops and an unnecessary if branch.

sriz1 replied:
addressed!

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+        self.energies.extend(excited_energies)
+
+
+    def V(self, X: ArrayLike) -> ArrayLike:
```
Why is this function necessary?

sriz1 replied:
Well if I don't use this function I get this error - and this is I think has to do with electronics.py - also it says "diabatic model" which is not true so maybe a typo? 

<img width="846" alt="Screen Shot 2022-04-07 at 4 50 49 PM" src="https://user-images.githubusercontent.com/46786940/162299593-98954ca2-3c96-41b9-8ae3-0c5e91573a27.png">

smparker replied:
Oh that's because you're deriving from `AdiabaticModel_` when you should be just driving from the `ElectronicsModel_` base class. `AdiabaticModel_` is for used for things like the Shin-Metiu model.

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+
+        V = self.V(X)
+
+        self.reference, self.hamiltonian = self._compute_basis_states(self.V(X), reference=reference)
```
I'm not sure this should be there at all...

sriz1 replied:
same issue as above

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+        """
+        self.call_and_parse_turbomole()
+
+        V = self.V(X)
```
Unused?

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+
+        self.reference, self.hamiltonian = self._compute_basis_states(self.V(X), reference=reference)
+
+        dV = self.dV(X)
```
unused?

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+        dV = self.dV(X)
+
+
+    def dV(self, X: ArrayLike) -> ArrayLike:
```
I don't think this should exist for this class

sriz1 replied:
resolved

smparker commented on mudslide/models.py:
```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+        return out
+
+
+    def cumSH(self):
```
This should have a more descriptive name, like `clone` or something.

sriz1 replied:
used clone

smparker commented on test/test_turbomole.py:
```diff
@@ -0,0 +1,231 @@
+
+        traj = mudslide.TrajectorySH(model, positions, mom, 3, tracer = YAMLTrace(), dt = 20, max_time = 41, t0 = 1) 
+        results = traj.simulate()
+
```
Remove the excess space.

smparker commented on test/test_turbomole.py:
```diff
@@ -0,0 +1,231 @@
+        self.assertAlmostEqual(excited_1_energy, ex_e_1_from_egrad, places=10)
+        self.assertAlmostEqual(excited_2_energy, ex_e_2_from_egrad, places=10)
+
+        np.testing.assert_almost_equal(gs_gradients, gs_grad_from_rdgrad,decimal = 10)
```
Do these `np.testing` functions work with `pytest` the way you're expecting? I have no idea so I'm genuinely asking.

smparker commented on test/test_turbomole.py:
```diff
@@ -0,0 +1,231 @@
+        np.testing.assert_almost_equal(gs_gradients, gs_grad_from_rdgrad,decimal = 10)
+        np.testing.assert_almost_equal(ex_st_1_gradients, ex_st_1_gradients_from_egrad,decimal = 10)
+        np.testing.assert_almost_equal(ex_st_2_gradients, ex_st_2_gradients_from_egrad,decimal = 10)
+
```
Why so much excess space?

sriz1 replied:
removed extra space

smparker commented on test/test_turbomole.py:
```diff
@@ -0,0 +1,231 @@
+   -1.74402803906479,      0.00000000000000,     -2.32753092373037,      
+    ])
+
+        coord_from_control = coord
```
This isn't actually testing anything...

sriz1 replied:
you're right - deleted!

smparker commented on test/test_turbomole.py:
```diff
@@ -0,0 +1,231 @@
+
+
+
+        gs_energy =  -78.40037178008 
```
Name all these variables with something like `_ref` so it's easier to tell these are reference values

sriz1 replied:
done

smparker commented on test/test_turbomole.py:
```diff
@@ -0,0 +1,231 @@
+
+        np.testing.assert_almost_equal(gs_gradients, gs_grad_from_rdgrad,decimal = 10)
+        np.testing.assert_almost_equal(ex_st_1_gradients, ex_st_1_gradients_from_egrad,decimal = 10)
+        np.testing.assert_almost_equal(ex_st_2_gradients, ex_st_2_gradients_from_egrad,decimal = 10)
```
10 decimal places is a lot and may end up failing if your convergence criteria aren't super tight to match.

sriz1 replied:
used 6

smparker commented on test/test_turbomole.py:
```diff
@@ -0,0 +1,231 @@
+                         -0.13059544360067088, 0.050433947305599816, 0.9797130350180959, -1.3877787807814457e-17]])
+
+
+        with open("traj-0-log_0.yaml", "r") as f:
```
This is reading data that was computed previously by a separate test? I'm not sure you can guarantee the order that tests run in. It would be better to make this part of the same test that generated the data.

sriz1 replied:
fixed it now (?)

smparker commented on test/test_turbomole.py:
```diff
@@ -0,0 +1,231 @@
+        np.testing.assert_almost_equal(dm_t_3, dm_from_mudslide_t3,  decimal = 10)
+
+
+    def tearDown(self):
```
What is this supposed to be doing?

sriz1 replied:
I was removing the files at the end but I no longer use

smparker commented on test/test_turbomole.py:
```diff
@@ -0,0 +1,231 @@
+    def setUp(self):
+        self.turbomole_files = ["control"]
+        for fl in self.turbomole_files:
+            shutil.copy("turbomole_files/"+fl,".")
```
This `turbomole_files` directory does not appear anywhere on the repo.

sriz1 replied:
added now

smparker replied:
Right now this only works if you run `pytest` in the `test/` directory. But you should be able to run `pytest` from the base directory as well (or potentially anywhere). You should be able to accomplish that by having the code look for the location of the present test being run (which is different from the current location) and use that as a base directory.

smparker reviewed (changes requested):
Still some missing files from the branch, like the periodic table data.

sriz1 committed:
Update setup.py

Commit:
TMModel.py class is edited significantly addressing comments for stramlining the code and added periodic table.py and constants.py file

Commit:
adding original models.py without TMModel.py in it

Commit:
TMModel.py and periodic_table.py files are now added

Commit:
tweaking in arounf update_coord function in TMModel.py file

smparker commented on mudslide/constants.py:
```diff
@@ -3,3 +3,4 @@
 
 boltzmann = 3.166811563455557e-6
 eVtoHartree = 27.211386245988
+amu_to_au = 1822.88825
```
Where is this coming from? It's not the value I'm back calculating from NIST.

sriz1 replied:
edited - I use this site: 
<img width="328" alt="Screen Shot 2022-04-10 at 2 26 16 PM" src="https://user-images.githubusercontent.com/46786940/162634108-b0521046-7fb0-4e4e-ab02-d46fd0103672.png">

and here are my back calculations from NIST values.

<img width="1030" alt="Screen Shot 2022-04-10 at 2 28 02 PM" src="https://user-images.githubusercontent.com/46786940/162634122-fd52beb2-dd80-4ff9-8f0b-2cb4af43b827.png">

smparker replied:
Yeah, that's what I was finding.

smparker commented on mudslide/periodic_table.py:
```diff
@@ -0,0 +1,135 @@
+
+# masses of most common isotopes to 5 decimal points, from
+# http://physics.nist.gov/cgi-bin/Compositions/stand_alone.pl
+masses = {
```
Use all of the digits that the nist table provides. No reason to cut off at 5 decimal points.

sriz1 replied:
addressed

Commit:
update periodic_table.py, constants.py, and TMModel.py
get the atomic unit conversion factor from NIST data file constants.py
provide full precision atomic masses in periodic_table.py
stramlining in get and update coord section. ALso now call_Turbomole function works with other modules of Turbomole

Commit:
Merge branch 'turbomole-model' of github.com:smparker/mudslide into turbomole-model

Commit:
streamline TMModel.py
Also this can now use other modules of Turbomole  using call_Turbomole function

Commit:
Use full precision atomic masses form NIST data

Commit:
add test_turbomole.py file
it now can check the test including dynamics

Commit:
add test files
control file inside folder turbomole_files is now added

smparker commented on setup.py:
```diff
@@ -26,6 +26,7 @@ def readme():
         'scipy',
         'typing_extensions',
         'pyyaml'
```
Should be a comma at the end of this line

sriz1 replied:
fixed.

sriz1 committed:
Update setup.py

Commit:
change filename TMModel to turbomole_model

Commit:
change filename TMModel to turbomole_model

Commit:
import TMModel inside even_sampling.py correctly
now TMModel class inside even_sampling.py is imported correctly (line 15)

Commit:
Merge branch 'turbomole-model' of github.com:smparker/mudslide into turbomole-model
Add turbomole_model file to mudslide

smparker committed:
check if turbomole is installed, skip tests if not

Commit:
resolve bug in test_turbomole.py

Commit:
Merge branch 'turbomole-model' of github.com:smparker/mudslide into turbomole-model
fix error in test_turbomole.py file

Commit:
fix raise exception in trajectory_sh.py line 40 to 50

Commit:
delete unnecessary print statements

Commit:
replace ground with 0

Commit:
Remove TMModel.py

Commit:
Remove au_to_amu.py

smparker commented on mudslide/even_sampling.py:
```diff
@@ -145,6 +146,20 @@ def clone(self, spawn_stack: Optional[Any] = None) -> 'EvenSamplingTrajectory':
         if spawn_stack is None:
             spawn_stack = self.spawn_stack
 
+#
+#You should basically always have to returning a 
+#clone and a model.clone will return a copy which 
+#that also copies the resources it will need 
+#So you don’t want this to do anything special 
+#based on the specific model – you just want this 
+#here should always do a clone (in evensampling trajectory) 
+#and then you should make a clone behave sensibly for a 
+#TMModel and every other model. For all the diabatic model – 
+#you can make a base clone – all that does is the return of the copy of the current data  
+#
+        if isinstance(self.model, TMModel):
+            self.model.clone()
+        print("Srizinevensampling")
```
None of this is doing anything yet, so remove.

sriz1 replied:
removed.

smparker commented on mudslide/turbomole_model.py:
```diff
@@ -0,0 +1,234 @@
+
+
+    def update_coords(self, X):
+        self.coord_path = subprocess.run(["sdg", "-f", "coord"], capture_output=True, text=True).stdout.rstrip()
```
This should probably be done just once when the class is first initialized and then stored.

sriz1 replied:
moved to here (line 57) 


<img width="863" alt="Screen Shot 2022-04-13 at 4 37 45 PM" src="https://user-images.githubusercontent.com/46786940/163266266-8502cdee-8c15-4842-af09-014d58f19364.png">

smparker commented on mudslide/turbomole_model.py:
```diff
@@ -0,0 +1,234 @@
+
+        current_line = read_start +1
+        for i, coord_list in enumerate(X):
+            lines[current_line] = "    {: .14f}      {: .14f}      {: .14f}      {}\n".format(
```
Use the fixed width scientific number notation we talked about previously.

sriz1 replied:
<img width="756" alt="Screen Shot 2022-04-13 at 4 09 51 PM" src="https://user-images.githubusercontent.com/46786940/163262005-f22f4826-6cd4-4e3a-badb-75ac72a620e9.png">


modified as suggested  - but the coordinate looks like this - I will have to check if Turbomole read it like that? I will check tomorrow once the server is back. 

<img width="557" alt="Screen Shot 2022-04-13 at 4 10 26 PM" src="https://user-images.githubusercontent.com/46786940/163262134-6f2077bc-e959-4d1d-9875-990e5405ebfa.png">

sriz1 replied:
fixed now

smparker commented on mudslide/turbomole_model.py:
```diff
@@ -0,0 +1,234 @@
+        X = X.reshape((self.ndim() // 3, 3))
+        with open(self.coord_path, "r") as f:
+            lines = f.readlines()
+        regex = re.compile(r"\$coord\n|\$coord\s")
+        read_start = np.argwhere(np.array([bool(regex.match(line)) for line in lines]))[0][0]
```
Simplify this. It's wasteful and hard to read as is. It will basically check every single line to see if it starts with `$coord`. Will it work if the line is ` $coord`? or `$coord ! a comment`?

sriz1 replied:
Ok I have tried this now !

sriz1 replied:
<img width="748" alt="Screen Shot 2022-04-14 at 5 54 41 PM" src="https://user-images.githubusercontent.com/46786940/163482884-f537fcf6-2aab-44ee-87b7-21fda076c42c.png">

OK I think this is good!

smparker replied:
Everything fails pretty catastrophically if the line starts with ` $coord` instead of `$coord` (just adding one space in front). Turbomole, on the other hand, works just fine in that case.

sriz1 replied:
<img width="547" alt="Screen Shot 2022-04-19 at 1 21 14 PM" src="https://user-images.githubusercontent.com/46786940/164060149-3ae312e4-15a2-441e-8c8e-7fd716bec4c5.png">


line 97 should fix that !

smparker commented on mudslide/turbomole_model.py:
```diff
@@ -0,0 +1,234 @@
+            coord_file.write("".join(lines))            
+
+
+    def call_turbomole(self, outname="turbo.out"):
+        # Open file/run Turbomole
+        with open(outname, "w") as f:
+            for turbomole_module in self.turbomole_modules.values():
+                tur_output = subprocess.run(turbomole_module, stdout=f) 
+
+        # Parse results with Turboparse
+        with open(outname, "r") as f:
+            data_dict = turboparse.parse_turbo(f)
```
this isn't doing anything... Are you expecting it to?

sriz1 replied:
Hmm I think this part of the code was repeated - deleted!

smparker commented on mudslide/turbomole_model.py:
```diff
@@ -0,0 +1,234 @@
+
+        if "egrad" in self.turbomole_modules.values():
+            # egrad updates to energy
+            parsed_energies = data_dict["egrad"]["excited_state"][0]["energy"]
```
Does this line do anything?

sriz1 replied:
What if someone calls turbomole_model but doesn’t want to run egrad? The code wouldn’t currently run, but this allows for the possibility for someone else to do something else.  makes sense to you?

smparker replied:
I meant this one line in particular that defines the variable `parsed_energies`. The variable `parsed_energies` isn't used anywhere else in the file.

sriz1 replied:
you're right - removed line parsed_energies

smparker commented on mudslide/turbomole_model.py:
```diff
@@ -0,0 +1,234 @@
+
+        self.force = -(self.gradients) 
+
+    def V(self, X: ArrayLike) -> ArrayLike:
```
delete

smparker commented on mudslide/turbomole_model.py:
```diff
@@ -0,0 +1,234 @@
+        file locations.)
+        """
+        self.call_turbomole()
+        self.reference, self.hamiltonian = self._compute_basis_states(self.V(X), reference=reference)
```
this function isn't supposed to be used for this context

smparker commented on mudslide/turbomole_model.py:
```diff
@@ -0,0 +1,234 @@
+        out.compute(X, couplings=couplings, gradients=gradients, reference=self.reference)
+        return out
+
+    def _compute_basis_states(self, V: ArrayLike, reference: Any = None) -> Tuple[ArrayLike,ArrayLike]:
```
This is a function for diabatic models. It shouldn't be here.

smparker commented on mudslide/turbomole_model.py:
```diff
@@ -0,0 +1,234 @@
+    def clone(self):
+        this_sub_dir = self.sub_dir_stem + "_" + str(self.sub_dir_num)
+        p = Path(self.turbomole_dir)/this_sub_dir
+        p.mkdir(parents = True, exist_ok = True)
```
If the directory already exists, this will cause some really weird problems...

sriz1 replied:
For now I deleted this clone function because this is doing nothing right now.

smparker commented on test/test_turbomole.py:
```diff
@@ -0,0 +1,135 @@
+
+        with open("TMtrace-0-log_0.yaml", "r") as f:
+            data = yaml.safe_load(f) 
+            print("amindata")
+            print(data)
```
delete unnecessary print statements

sriz1 replied:
done

smparker commented on test/test_turbomole.py:
```diff
@@ -0,0 +1,135 @@
+                2.145175145340160000, 0.594918215579156000, 1.075977514428970000,
+                -2.269965412856570000,  0.495551832268249000,   1.487150300486560000]
+
+        traj = mudslide.TrajectorySH(model, positions, mom, 3, tracer = YAMLTrace(name = "TMtrace"), dt = 20, max_time = 21, t0 = 1) 
```
So this only does one step? Is that enough to make sure everything is working right? Put another way, do the tests fail if you deliberately introduce a minor bug?

Commit:
remove unnecessary stuff and adjust coord update section

Commit:
remove usused codes and priting statements

Commit:
delete functions that do not belong to this class

Commit:
clear formats and spaces

Commit:
add boltzmann_velocities

Commit:
Revert "Update setup.py"

This reverts commit a5b4ca660612576eb76e85c11d807fec5644de38.

revert a specific commit to get test file back

Commit:
Revert "Revert "Update setup.py""

This reverts commit 6c0523cd440c7eb8fa3076e79aa3a31eefe7655a.

revert update setup.py

Commit:
Revert "resolve bug in test_turbomole.py"

This reverts commit bcd67ccf0b2fd2ddf422996857b110eeb6973f42.

revert commit

Commit:
Revert "clear formats and spaces"

This reverts commit 4cc8a7121cc25ce6bfebc6e576a1b6377cc61dd9.

Commit:
revert test_turbomole.py

Commit:
fix assertion error in test

Commit:
fix bug
also added test for position

Commit:
add update function back

smparker commented on mudslide/turbomole_model.py:
```diff
@@ -0,0 +1,192 @@
+                j = dct["ket_state"]
+
+                self.derivative_coupling[i][j] = np.array(dct["d/dR"]).reshape(self.ndim(), order="F")
+                self.derivative_coupling[j][i] = self.derivative_coupling[i][j]
```
This line is wrong. Derivative couplings are anti-Hermitian, not Hermitian. Meaning tau_ij = -tau_ji.

sriz1 replied:
should have been checking this  - corrected now! 

<img width="1034" alt="Screen Shot 2022-04-19 at 2 06 38 PM" src="https://user-images.githubusercontent.com/46786940/164067915-80eb2f81-f708-485c-b830-a652d452b126.png">

smparker commented on mudslide/batch.py:
```diff
@@ -17,6 +17,21 @@
 logger = logging.getLogger("mudslide")
 
+
+def boltzmann_velocities(mass, temperature, scale=True, seed=None): 
+    rng = np.random.default_rng(seed)
+    kt = boltzmann * temperature
+    sigma = np.sqrt(kt * mass)
+    p = rng.normal(0.0, sigma)
+    if scale:
+        avg_KE = 0.5 * np.dot(p**2, np.reciprocal(mass))/mass.size
+        kbT2 = 0.5 * kt
+        scal = np.sqrt(kbT2/avg_KE)
+        p *= scal
+    return p
```
- I don't think this is the best place for this code. Maybe it'd be better in a new file for physical manipulations? Or `math.py`? Or `util.py`? I'm not sure, but it's out of place in `batch.py`
  - Add a test that covers this function.

Commit:
fix sign in derivative coupling

Commit:
add testdir
pytest can be run for files including test_turbomole.py  from anywhwere in the mudslide

Commit:
change grid m3 to 3

smparker committed:
reload yaml traces from any location

- simplify tests so they don't dump yaml files everywhere
- test yaml logging directly

Commit:
use exact same max_time

Commit:
use same max_time

Commit:
fix the mass issues
reference values were using Turbomole based masses for C and H

smparker committed:
fix indexing in yaml logs

smparker committed:
restart trajectories from yaml

smparker committed:
fix testdir usage

smparker committed:
run tm test in checkdir

smparker committed:
make sdg() a class function

smparker committed:
check then discard stderr

smparker committed:
remove trailing whitespace

smparker committed:
apply recommended turbomole settings by default

smparker committed:
remove turbomole_files data in favor of ref data

Commit:
postprocess gradients in a right way

"postprocess gradients in a right way
conflict resolved"

Commit:
fix conflict

Commit:
remove debugging print statements

Commit:
revove print statements

Commit:
clean the test file

Commit:
add referece file

Commit:
add referece file

Commit:
add referece file

smparker committed:
require turboparse 0.1.2

Commit:
remove boltzmann_velocity function

Commit:
return velocity

Commit:
cleanup test

Commit:
add test for boltzmann function

Commit:
clean test_turbomole  further

Commit:
remove turbomole call in boltzmann test

smparker commented on test/test_boltzmann_velocity.py:
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""Unit testing for Turbomole_class"""
```
update title

smparker commented on test/test_boltzmann_velocity.py:
```diff
@@ -0,0 +1,41 @@
+
+import numpy as np
+import os
+import shutil
```
remove unused imports

smparker commented on mudslide/math.py:
```diff
@@ -5,10 +5,25 @@
+        kbT2 = 0.5 * kt
+        scal = np.sqrt(kbT2/avg_KE)
+        p *= scal
+        p = p / mass
```
unindent this line. It should turn the p into velocities even if scale is False

Commit:
update title and remove unused imports

Commit:
unindent a line for getting velocities

smparker committed:
remove added white space

smparker commented on mudslide/electronics.py:
```diff
@@ -64,6 +65,11 @@ def as_dict(self):
+
+##    def clone(self):
+##        return self
+        
```
```suggestion
        
```

smparker commented on mudslide/electronics.py:
```diff
@@ -98,6 +104,8 @@ def update(self, X: ArrayLike, couplings: Any = None, gradients: Any = None) ->
         return out
 
+
+
```
```suggestion

```

smparker commented on mudslide/electronics.py:
```diff
@@ -155,13 +163,14 @@ def _compute_derivative_coupling(self, coeff: ArrayLike, dV: ArrayLike, energies
         return out
 
+
+
```
```suggestion
```

smparker commented on mudslide/even_sampling.py:
```diff
@@ -12,6 +12,7 @@
 
 from typing import Optional, List, Any, Dict, Union
 from .typing import ArrayLike, ElectronicT
+from .turbomole_model import TMModel
```
```suggestion
```

smparker committed:
remove commented code

smparker committed:
clean whitespace

smparker committed:
clean whitespace

smparker committed:
remove unnecessary import

smparker committed:
put all models under mudslide.models
